packages/transport-webtransport/src/index.ts

```diff
@@ -52,60 +64,122 @@ class WebTransportTransport implements Transport {
-    }
+    let abortListener: (() => void) | undefined
+    let maConn: MultiaddrConnection | undefined
 
-    wt.closed.catch((err: Error) => {
-      log.error('WebTransport connection closed:', err)
-      // This is how we specify the connection is closed and shouldn't be used.
-      maConn.timeline.close = Date.now()
-    })
+    let cleanUpWTSession: (closeInfo?: WebTransportCloseInfo) => void = () => {}
 
     try {
+      let closed = false
+      const wt = new WebTransport(`${url}/.well-known/libp2p-webtransport?type=noise`, {
+        serverCertificateHashes: certhashes.map(certhash => ({
+          algorithm: 'sha-256',
+          value: certhash.digest
+        }))
+      })
+
+      cleanUpWTSession = (closeInfo?: WebTransportCloseInfo) => {
+        try {
+          if (maConn != null) {
+            if (maConn.timeline.close != null) {
+              // already closed session
+              return
+            }
+
+            // This is how we specify the connection is closed and shouldn't be used.
+            maConn.timeline.close = Date.now()
+          }
+
+          if (closed) {
+            // already closed session
+            return
+          }
+
+          wt.close(closeInfo)
+        } catch (err) {
+          log.error('error closing wt session', err)
+        } finally {
+          closed = true
+        }
+      }
+
+      // this promise resolves/throws when the session is closed or failed to init
+      wt.closed
+        .then(async () => {
+          await maConn?.close()
+        })
+        .catch((err: Error) => {
+          log.error('error on remote wt session close', err)
+          maConn?.abort(err)
+        })
+        .finally(() => {
+          // if we never got as far as creating the maConn, just clean up the session
+          if (maConn == null) {
+            cleanUpWTSession()
+          }
+        })
+
+      // if the dial is aborted before we are ready, close the WebTransport session
+      abortListener = () => {
+        if (abortListener != null) {
+          options.signal?.removeEventListener('abort', abortListener)
+        }
+
+        cleanUpWTSession()
+      }
+      options.signal?.addEventListener('abort', abortListener)
+
+      await wt.ready
+
+      if (!await this.authenticateWebTransport(wt, localPeer, remotePeer, certhashes)) {
+        throw new Error('Failed to authenticate webtransport')
+      }
+
+      maConn = {
+        close: async (options?: AbortOptions) => {
+          log('Closing webtransport')
+          cleanUpWTSession()
+        },
+        abort: (err: Error) => {
+          log('aborting webtransport due to passed err', err)
+          cleanUpWTSession({
+            closeCode: 0,
+            reason: err.message
+          })
+        },
+        remoteAddr: ma,
+        timeline: {
+          open: Date.now()
+        },
+        // This connection is never used directly since webtransport supports native streams.
+        ...inertDuplex()
+      }
```

Comment:
if we make this a property on the class we can split out this dial function a bit